libarchive/archive_write_set_format_mtree.c

```diff
@@ -1754,134 +1753,133 @@ mtree_entry_setup_filenames(struct archive_write *a, struct mtree_entry *file,
 #else
 	(void)a; /* UNUSED */
 #endif
-	pathname =  file->pathname.s;
+	pathname = file->pathname.s;
 	if (strcmp(pathname, ".") == 0) {
 		archive_strcpy(&file->basename, ".");
 		return (ARCHIVE_OK);
 	}
 
-	archive_strcpy(&(file->parentdir), pathname);
-
-	len = file->parentdir.length;
-	p = dirname = file->parentdir.s;
-
-	/*
-	 * Remove leading '/' and '../' elements
-	 */
-	while (*p) {
-		if (p[0] == '/') {
-			p++;
+	/* Remove leading '/' and '../' elements */
+	len = archive_strlen(&file->pathname);
+	while (*pathname) {
+		if (pathname[0] == '/') {
+			pathname++;
 			len--;
-		} else if (p[0] != '.')
+		} else if (pathname[0] != '.') {
 			break;
-		else if (p[1] == '.' && p[2] == '/') {
-			p += 3;
+		} else if (pathname[1] == '.' && pathname[2] == '/') {
+			pathname += 3;
 			len -= 3;
-		} else
+		} else {
 			break;
+		}
 	}
-	if (p != dirname) {
-		memmove(dirname, p, len+1);
-		p = dirname;
+
+	/*
+	 * Ensure that we have enough room for the './' prefix
+	 * and '\0' terminator.
+	 */
+	archive_string_ensure(&file->parentdir, len + 2 + 1);
+
+	/*
+	 * Add "./" prefix.
+	 * NOTE: If the pathname does not have a path separator, we have
+	 * to add "./" to the head of the pathename because mtree reader
+	 * will suppose that it is v1(a.k.a classic) mtree format and
+	 * change the directory unexpectedly and so it will make a wrong
+	 * path.
+	 */
+	if (strncmp(pathname, "./", 2) != 0)
+		archive_strncat(&file->parentdir, "./", 2);
```

Comment:
Technically parentdir is empty there, so it can be considered as a prefix or suffix. Wording is at your convenience :)